treewide: removed all usages of seastar io priority classes

mmaslankaprv opened the pull request:
Removed all references to `seastar::io_priority_class`. The io priority class was removed from seastar together with introduction of API version 7. The use of the priority class was deprecated. Now the `io_priority_class` was completely removed. This PR adapts the usage of Seastar APIs not to use io priority classes.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
treewide: removed all usages of seastar io priority classes

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--240032b7-7aea-4d7c-a9e2-8cc07412d112-->
<!--TEST RESULTS-->
<!--7c53295d-4cbf-4ff3-bd66-92947a173a9b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66406">build#66406</a></summary>


| test_class                                   | test_method                                  | test_arguments                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------------------|:---------------------------------------------|:----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AccessControlListTestUpgrade                 | test_upgrade_sasl                            |                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66406#01970cc2-f3e6-47d1-8fe9-a0adc7f9783e | FLAKY         | 20/21    | upstream reliability is '97.41935483870968'. current run reliability is '95.23809523809523'. drift is 2.18126 and the allowed drift is set to 50. The test should PASS |
| ConsumerOffsetsRecoveryToolTest              | test_consumer_offsets_partition_count_change |                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66406#01970cc2-f3e6-47d1-8fe9-a0adc7f9783e | FLAKY         | 20/21    | upstream reliability is '97.84946236559139'. current run reliability is '95.23809523809523'. drift is 2.61137 and the allowed drift is set to 50. The test should PASS |
| DataTransformsTest                           | test_identity                                | {"transactional": true, "wait_running": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66406#01970cc2-f3e6-47d1-8fe9-a0adc7f9783e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| DisablingPartitionsTest                      | test_disable                                 |                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66406#01970cc2-f3e7-4e39-bfe0-772c2aab42d9 | FLAKY         | 16/21    | upstream reliability is '78.41614906832298'. current run reliability is '76.19047619047619'. drift is 2.22567 and the allowed drift is set to 50. The test should PASS |
| src/v/datalake/tests/record_multiplexer_test | src/v/datalake/tests/record_multiplexer_test |                                               | unit        | https://buildkite.com/redpanda/redpanda/builds/66406#01970c6a-dbe5-4aee-8b57-3d9b7ac2a7ff | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66411">build#66411</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:---------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66411#01970df2-f7b5-4a7c-87b4-5bce7eefc698 | FLAKY         | 20/21    | upstream reliability is '95.08670520231213'. current run reliability is '95.23809523809523'. drift is -0.15139 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66411#01970df2-f7b5-4a7c-87b4-5bce7eefc698 | FLAKY         | 20/21    | upstream reliability is '95.50561797752809'. current run reliability is '95.23809523809523'. drift is 0.26752 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

mmaslankaprv committed:
c/controller: pass `cluster` scheduling group into controller

Controller services are supposed to be scheduled within a `cluster`
scheduling group. Passed the scheduling group to allow controller
services access to it.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
c/controller_stm: use `cluster` scheduling group in controller stm

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
c/controller_backed: use `cluster` scheduling group in ctrl backend

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
c/controller: switch to `cluster` scheduling group when starting

Switching to `cluster` scheduling group when starting all services
aggregated within the `controller` will make all the reconciliation
loops run in the `cluster` scheduling group.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

mmaslankaprv commented:
/dt

StephanDollberg commented on src/v/cloud_io/io_resources.cc:
```diff
@@ -263,19 +263,17 @@ ss::future<> io_resources::set_disk_max_bandwidth(size_t tput) {
-          .shadow_indexing_priority()
-          .update_bandwidth(tput);
+        // TODO: Revisit when we update seastar.
+        // co_await _scheduling_group.update_io_bandwidth(tput);
```
This should still work (just maybe different API)?

StephanDollberg commented on src/v/resource_mgmt/io_priority.h:
```diff
@@ -1,105 +0,0 @@
-private:
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"
-    priority_manager()
```
Did you check that these match (mostly)?

mmaslankaprv replied:
they mostly match, at least in the critical paths.

StephanDollberg commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -2287,14 +2281,11 @@ ss::future<> disk_log_impl::apply_segment_ms() {
         co_return;
     }
 
-    auto pc = last->appender()
```
any idea why we did this extra hop?

mmaslankaprv replied:
it looks that we only stored the priority information in `segment_appender` this way the new appender was created with the priority of the previous one

StephanDollberg commented on src/v/storage/file_sanitizer.h:
```diff
@@ -105,15 +105,15 @@ class file_io_sanitizer final : public ss::file_impl {
       const void* buffer,
       size_t len,
-      const ss::io_priority_class& pc) final {
+      ss::io_intent* intent = nullptr) final {
```
Why are we introducing `io_intent` (they are kinda weird) here?

mmaslankaprv replied:
Sanitizer implements the `ss::file` interface which uses the intents. I agree they are weird: 
```
    virtual future<size_t> write_dma(uint64_t pos, const void* buffer, size_t len, io_intent*) = 0;
    virtual future<size_t> write_dma(uint64_t pos, std::vector<iovec> iov, io_intent*) = 0;
    virtual future<size_t> read_dma(uint64_t pos, void* buffer, size_t len, io_intent*) = 0;
    virtual future<size_t> read_dma(uint64_t pos, std::vector<iovec> iov, io_intent*) = 0;
    virtual future<temporary_buffer<uint8_t>> dma_read_bulk(uint64_t offset, size_t range_size, io_intent*) = 0;
```

StephanDollberg commented on src/v/test_utils/tmpbuf_file.h:
```diff
@@ -55,17 +55,15 @@ class tmpbuf_file final : public ss::file_impl {
       const void* buffer,
       const size_t len,
-      const ss::io_priority_class&) final {
+      ss::io_intent*) final {
```
same question as above about io_intent

mmaslankaprv replied:
the same, `ss::file` interface

StephanDollberg commented on src/v/cluster/controller.cc:
```diff
@@ -823,6 +829,8 @@ ss::future<> controller::start(
     co_await _data_migration_irpc_frontend.start(
       std::ref(_feature_table), std::ref(_data_migration_backend));
+    // switch back to the default scheduling group
+    co_await ss::coroutine::switch_to(ss::default_scheduling_group());
```
Are we sure this is the correct SG? Shouldn't we rather do the explicit restore previous one.

rockwotj replied:
This is a noop, switch_to has RAII like semantics that it only takes effect within the current coroutine scope.

But plus +1 to Stephan's comment about previous one, but you get that for free by just deleting this line. If you did want to explicitly switch back, switch_to returns the previous group (but again you don't need that here).
https://github.com/scylladb/seastar/blob/8df8212e53577e1d8477a5c901457cd61d88afc7/tests/unit/coroutines_test.cc#L249

StephanDollberg commented:
thanks for doing this

rockwotj commented on bazel/thirdparty/seastar.BUILD:
```diff
@@ -65,7 +65,7 @@ bool_flag(
 int_flag(
     name = "api_level",
-    build_setting_default = 6,
+    build_setting_default = 7,
```
I ❤️ we don't have to do a vtools dance for this kind of change anymore

rockwotj reviewed (commented):
Amazing. Thanks for slogging through this

mmaslankaprv committed:
c_storage/tests: made timeout for segment_chunk_hydration test longer

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/ci-repeat 1

mmaslankaprv commented:
/ci-repeat 1

piyushredpanda commented:
@rockwotj or @StephanDollberg: do either of you think this is good to approve + merge?

StephanDollberg commented:
@mmaslankaprv actually one thing I notice now is that we no longer have a separate TS downloads group now, right? It all runs in `fetch` now? Previously we had `shadow-indexing` with 500 weight.

In a way this was the most important group as that is the only one that can really produce massive amounts of write throughput. 

FYI there is currently this discussion going on about io-properties https://redpandadata.atlassian.net/browse/CORE-10020 which kinda related to this topic.

mmaslankaprv commented:
going to force merge it as the failures are not related
